opentelemetry/src/sdk/trace/provider.rs

```diff
@@ -145,17 +145,15 @@ impl Builder {
         let mut config = self.config;
         let sdk_provided_resource = Resource::from_detectors(
             Duration::from_secs(0),
-            vec![Box::new(SdkProvidedResourceDetector)],
+            vec![
+                Box::new(SdkProvidedResourceDetector),
+                Box::new(EnvResourceDetector::new()),
+            ],
         );
         config.resource = match config.resource {
             None => Some(Arc::new(sdk_provided_resource)),
-            Some(resource) => {
-                if resource.is_empty() {
-                    None
-                } else {
-                    Some(resource)
-                }
-            }
+            // User provided resource information has higher priority.
+            Some(resource) => Some(Arc::new(sdk_provided_resource.merge(resource))),
```

Comment:
I think we can keep the `merge` function as it is and switch the `resource` and `sdk_provided_resource`. i. e `resource.merge(sdk_provided_resource)`?

Reply:
I don't think we could do that since it wouldn't be spec complaint?

Reply:
Either way we should check the benchmarks to see if we still want the optimization for the case where both are empty.

Reply:
Will both of them ever be empty? At least there will be SDK provided [default values](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/README.md#semantic-attributes-with-sdk-provided-default-value), right?

Reply:
Currently the sdk allows users who are focused on the lowest overhead to out through specifying [with_no_resource](https://docs.rs/opentelemetry/0.16.0/opentelemetry/sdk/trace/struct.Config.html#method.with_no_resource).

Reply:
Just want to make sure I am understanding you clearly. When you say _both are empty_ What are you referring to here? I can see how `config.resource` can be empty, but what is there other one?

Reply:
Ah yeah just the `config.resource` check that was there before.

Reply:
The current code vs something like code below?

```rust
if resource.is_empty() {
	Some(Arc::new(sdk_provided_resource))
} else {
	Some(Arc::new(sdk_provided_resource.merge(resource))),
}
```
You wanted to check the benchmarks when that resource provided empty, right?

Reply:
The current optimization is to have the resource be `None` in this case.

Reply:
Hmm, that would not be conforming to the specification. I think it would be surprising for users to find this behaviour different in rust SDK if they were also using other language SDKs. If you prefer that way I can change it back to `None`.

Reply:
I think it would be simpler to keep the current optimization and have a subsequent issue/pr around removing it as `with_no_resource` would also need to be removed.

Reply:
Sure, that sounds good to me.